dali/operators/reader/coco_reader_op.cc

```diff
@@ -119,11 +123,177 @@ void COCOReader::ValidateOptions(const OpSpec &spec) {
       "When reading data from meta files `dump_meta_files_path` option is not supported.");
   }
 
+  if (spec.HasArgument("masks")) {
+    DALI_ENFORCE(
+      !spec.HasArgument("pixelwise_masks"),
+      "`masks` and `pixelwise_masks` cannot be both provided.");
+  }
+
   if (spec.HasArgument("dump_meta_files")) {
     DALI_ENFORCE(
       spec.HasArgument("dump_meta_files_path"),
       "When dumping meta files `dump_meta_files_path` must be provided.");
   }
 }
 
+void COCOReader::PixelwiseMasks(int image_id, int* mask) {
+  const auto &meta = masks_meta_[image_id];
+  const auto &coords = mask_coords_[image_id];
+  int h = heights_[image_id];
+  int w = widths_[image_id];
+  const int *labels_in = labels_.data() + offsets_[image_id];
+  int labels_size = counts_[image_id];
+  std::set<int> labels(labels_in, labels_in + labels_size);
+  if (! labels.size()) {
+    return;
+  }
+
+  // Create a run-length encoding for earch polygon, indexed by label :
+  std::map<int, std::vector<RLE> > frPoly;
+  for (uint polygon_idx = 0; polygon_idx < meta.size() / 3; polygon_idx++) {
+    int mask_idx = meta[3 * polygon_idx];
+    int start_idx = meta[3 * polygon_idx + 1];
+    int end_idx = meta[3 * polygon_idx + 2];
+    int label = *(labels_in + mask_idx);
+    // Convert polygon to encoded mask
+    double* in = new double[end_idx - start_idx];
+    for (int i = 0; i < end_idx - start_idx; i++)
+      in[i] = (double) coords[start_idx + i];
+    RLE M;
+    rleInit(&M, 0, 0, 0, 0);
+    rleFrPoly(&M, in, (end_idx - start_idx) / 2, h, w);
+    if (frPoly.find(label) == frPoly.end()) {
+      std::vector<RLE> v;
+      v.push_back(M);
+      frPoly.insert(std::make_pair(label, v));
+    } else {
+      frPoly[label].push_back(M);
+    }
+    delete in;
+  }
+
+  // Create a run-length encoding for each compressed string representation
+  RLE* frString;
+  rlesInit(&frString, masks_rles_[image_id].size());
+  uint ann_cnt = 0;
+  for (auto str : masks_rles_[image_id])
+    rleFrString(&frString[ann_cnt++], (char *) str.c_str(), h, w);
+
+  // Reserve run-length encodings by labels
+  RLE* R;
+  rlesInit(&R, *labels.rbegin() + 1);
+
+  // Merge each label (from polygons annotations)
+  uint lab_cnt = 0;
+  for (auto rles : frPoly)
+    for (auto rle : rles.second)
+      rleMerge(&rle, &R[rles.first], 1, 0);
+  // Merge each label (from RLE annotations)
+  ann_cnt = 0;
+  for (auto mask_idx : masks_rles_idx_[image_id]) {
+    int label = *(labels_in + mask_idx);
+    rleMerge(&frString[ann_cnt++], &R[label], 1, 0);
```

Comment:
It can't be done in place.
This code again is incorrect, but I think it's just useless - run-length encodings are meant to cover the whole label by design, there's no need to merge anything here. I'll remove it.